### DISCUSSION TITLE
masterless mode?

### DISCUSSION THREAD

**Discussion by lollipopman** [choria-io/general#25] [Q&A] [answered]
Does choria support a masterless puppet mode, something akin to:

```
$ git checkout git@github.com/example/puppet.git
$ choria puppet apply node.example.com
```

In other words choria would copy the files from the current directory and apply them on the destination host using puppet. Two other projects which support this mode:

- [Puppet Bolt](https://github.com/puppetlabs/bolt)
- [supply_drop](https://github.com/pitluga/supply_drop)

**Comment by ripienaar** [comment, marked as answer]
Nope. I have had ssh based transports in the past(though they didn’t copy themselves or payload like ansible) but I think it’s just not that kind of tool ultimately.

**Reply by lollipopman** [reply]
Thanks @ripienaar, I'm curious, what is your puppet development workflow like without a masterless option? I commonly use masterless noops to check whether my code is working correctly against a specific server, without having to cut a commit.

**Reply by ripienaar** [reply]
Puppet development against unit tests using rspec puppet and with its quite mature integration tool set largely removes the need for that kind of thing - though I agree its quite an undertaking to get to that.

However, it's worth noting that you can solve this problem yourself. You could write an agent that does your steps - git clone and call puppet apply - and just invoke that agent using Choria.

Largely Choria is a framework for building this kind of thing - we ship very little by way of baked end user ready things, the puppet module being the nearest I guess - but it has a multitude of extension points allowing you to create Agent plugins that does anything you want and exposing those over the RPC layer either for programatic access or view like `choria req` etc.

**Reply by lollipopman** [reply]
Thanks for the detailed reply @ripienaar, I'll try to find some time to explore choria more in depth and see if I can create a masterless prototype using choria's rpc framework.

**Reply by ripienaar** [reply]
Yeah, what I thought you were asking about was replacing the choria broker communications with something else - that's a no - but for making it do things like pull code and run puppet, no biggie that you can do with agents.
